common/distribution/dist.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2016 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package distribution
+
+import (
+	"fmt"
+	"net/url"
+)
+
+// DistType represents a distribution type
+type DistType string
+
+const (
+	DistScheme = "cimd"
+
+	// Distribution types
+	DistTypeAppc       DistType = "appc"
+	DistTypeACIArchive DistType = "aci-archive"
+	DistTypeDocker     DistType = "docker"
+)
+
+// A Distribution represnt the way to retrieve an image starting from an input
+// string.
+// It's identified by an URI with a specific schema
+type Distribution interface {
+	// Returns a copy of the distribution URI
+	URI() *url.URL
+	// Returns the distribution type
+	Type() DistType
```

Comment:
> I am hesitant to use DistType switches in upper architectual layers (although I am aware this is hard).

I'd also like to avoid this if possible. Some thoughts:

* Upper code needs a way to know the distribution type. e.g. for calling the right distribution handler. Possible solutions:

 * Keep the `Type() DistType` func from the Distribution interface as now.
 * Use an helper function `DistType(d Distribution) (DistType, error)` (removing the `Type() DistType` func from the Distribution interface)

* To avoid type assertions I can remove all the additional (not covered by the Distribution interface) methods from the distributions (`Appc.App() *discovery.App` for Appc, `ACIArchive.ArchiveURL() *url.URL` for ACIArchive, `Docker.DockerString()` `Docker.SimpleDockerString() string` for Docker) and implement generic helpers methods like 
```
AppDiscoveryString(d Distribution) (*discovery.App,error) // Will work only for Appc Distribution
TransportURL(d Distribution) (*url.URL, error) // Will work only for ACIArchive and perhaps we can use the same function for future distributions like oci-image-layout
DockerString(d Distribution) (string,error) // Will work only for Docker Distribution
SimpleDockerString(d Distribution) (string,error) // Will work only for Docker Distribution
```
Since they will have a Distribution interface as input (to avoid type assertions) they have to do validation since they'll work only for a specific distribution type (so they will also return an error) (and quite probably they'll do an internal type assertion).

This will avoid the need for the user to know the additional provided methods of a distribution type but just use some helper methods.

Thoughts? Better Ideas?